File: python/merge_sorted_array.py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        #ptr1 will traverse nums1 in reverse.  ptr2 will traverse nums2 in reverse
        ptr1, ptr2 = m - 1, n - 1
        #i traverses the full nums1 in reverse, adding the final elements
        for i in range(m + n - 1, -1, -1):
            if ptr1 < 0 or (ptr2 >= 0 and nums2[ptr2] > nums1[ptr1]):
                nums1[i] = nums2[ptr2]
                ptr2 -= 1
            else:
                nums1[i] = nums1[ptr1]
                ptr1 -= 1

File: java/mergeSortedArray.java
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        //pointer1 will traverse nums1 in reverse
        int pointer1 = m - 1;
        //pointer2 will traverse nums2 in reverse
        int pointer2 = n - 1;
        //i traverses the full nums1 in reverse
        for (int i = m + n - 1; i >= 0; i--) {
            if (pointer1 < 0 || (pointer2 >= 0 && nums2[pointer2] > nums1[pointer1])) {
                nums1[i] = nums2[pointer2];
                pointer2--;
            } else {
                nums1[i] = nums1[pointer1];
                pointer1--;
            }
        }
    }
}